students/john_navitsky/session04/trigrams.py

```diff
@@ -3,10 +3,43 @@
 import random
 
 #text = "I wish I may I wish I might"
-text = "One night--it was on the twentieth of March, 1888--I was returning from a journey to a patient (for I had now returned to civil practice), when my way led me through Baker Street. As I passed the well-remembered door, which must always be associated in my mind with my wooing, and with the dark incidents of the Study in Scarlet, I was seized with a keen desire to see Holmes again, and to know how he was employing his extraordinary powers. His rooms were brilliantly lit, and, even as I looked up, I saw his tall, spare figure pass twice in a dark silhouette against the blind. He was pacing the room swiftly, eagerly, with his head sunk upon his chest and his hands clasped behind him. To me, who knew his every mood and habit, his attitude and manner told their own story. He was at work again. He had risen out of his drug-created dreams and was hot upon the scent of some new problem. I rang the bell and was shown up to the chamber which had formerly been in part my own."
+#text = "One night--it was on the twentieth of March, 1888--I was returning from a journey to a patient (for I had now returned to civil practice), when my way led me through Baker Street. As I passed the well-remembered door, which must always be associated in my mind with my wooing, and with the dark incidents of the Study in Scarlet, I was seized with a keen desire to see Holmes again, and to know how he was employing his extraordinary powers. His rooms were brilliantly lit, and, even as I looked up, I saw his tall, spare figure pass twice in a dark silhouette against the blind. He was pacing the room swiftly, eagerly, with his head sunk upon his chest and his hands clasped behind him. To me, who knew his every mood and habit, his attitude and manner told their own story. He was at work again. He had risen out of his drug-created dreams and was hot upon the scent of some new problem. I rang the bell and was shown up to the chamber which had formerly been in part my own."
 
+input_file = open("boscombe_valley_mystery.txt","r")
+text = input_file.read()
+input_file.close()
+
+#print("RAW_TEXT:",text)
+
+# strip out most punctuation
+text=text.replace("--", " ")
+text=text.replace("-", " ")
+text=text.replace(":", "")
+text=text.replace("(", "")
+text=text.replace(")", "")
+text=text.replace("\"", "")
+text=text.replace("\'", "")
+
+# make a special placeholder for paragraphs
+text=text.replace("\n\n", " __paragraph__ ")
+
+# but make some punctuation into their own word
+text=text.replace(".", " .")
+text=text.replace(",", " ,")
+text=text.replace("?", " ?")
+text=text.replace("!", " !")
+
+# build the input list
 word_list = text.split()
-print("INPUT_LIST:",word_list)
+
+# lower case words following periods because they 
+# are probably not proper nouns
+for index, word in enumerate(word_list):
+	if word == "." and index < len(word_list)-1:
```

Comment:
you could do:

`for index, word enumerate(word_list[:-1]:`

to avoid the special logic.

or catch the IndexError.